File: go/validator.go
package securekit

import (
	"errors"
	"html"
	"net/mail"
	"net/url"
	"regexp"
	"strings"
)

// ErrInvalidInput is returned by validators when input fails validation.
var ErrInvalidInput = errors.New("securekit: invalid input")

// ValidateEmail reports whether email is a syntactically valid address
// (RFC 5322, via net/mail) and rejects surrounding whitespace/control chars.
func ValidateEmail(email string) bool {
	if email == "" || strings.ContainsAny(email, "\r\n\t") {
		return false
	}
	addr, err := mail.ParseAddress(email)
	if err != nil {
		return false
	}
	// mail.ParseAddress accepts "Name <addr>"; require the parsed address to
	// equal the raw input so display-name tricks are rejected.
	return addr.Address == email
}

// ValidateURL reports whether rawURL is an absolute http(s) URL. Other
// schemes (javascript:, data:, file:, etc.) are rejected to reduce
// SSRF/XSS-via-redirect risk in typical web use cases.
func ValidateURL(rawURL string) bool {
	u, err := url.ParseRequestURI(rawURL)
	if err != nil {
		return false
	}
	if u.Scheme != "http" && u.Scheme != "https" {
		return false
	}
	return u.Host != ""
}

var allowListCache = map[string]*regexp.Regexp{}

// ValidateAllowList reports whether s contains only characters present in
// allowedChars (a literal character set, e.g. "a-zA-Z0-9_-").
func ValidateAllowList(s, allowedChars string) bool {
	re, ok := allowListCache[allowedChars]
	if !ok {
		re = regexp.MustCompile(`^[` + allowedChars + `]*$`)
		allowListCache[allowedChars] = re
	}
	return re.MatchString(s)
}

// SanitizeFilename guards against path traversal by stripping directory
// components and rejecting empty/dot-only results. It returns the base
// filename only - callers must still join it with a trusted, fixed base
// directory rather than trusting any caller-supplied prefix.
func SanitizeFilename(filename string) (string, error) {
	if filename == "" {
		return "", ErrInvalidInput
	}
	// Reject NUL and path separators outright rather than silently stripping,
	// since silent stripping can turn "a/../b" into a different plausible name.
	if strings.ContainsAny(filename, "\x00/\\") {
		return "", ErrInvalidInput
	}
	trimmed := strings.TrimSpace(filename)
	if trimmed == "" || trimmed == "." || trimmed == ".." {
		return "", ErrInvalidInput
	}
	return trimmed, nil
}

// EscapeHTML output-encodes s for safe inclusion in HTML body context,
// preventing XSS. Equivalent to PHP's htmlspecialchars(ENT_QUOTES).
func EscapeHTML(s string) string {
	return html.EscapeString(s)
}

File: php/src/Validator.php
<?php

declare(strict_types=1);

namespace SecureKit;

/**
 * Input validation and output-encoding helpers.
 */
final class Validator
{
    public static function isValidEmail(string $email): bool
    {
        if ($email === '' || preg_match('/[\r\n\t]/', $email)) {
            return false;
        }
        return filter_var($email, FILTER_VALIDATE_EMAIL) !== false;
    }

    /**
     * Only absolute http(s) URLs are accepted; other schemes
     * (javascript:, data:, file:, ...) are rejected.
     */
    public static function isValidUrl(string $url): bool
    {
        if (filter_var($url, FILTER_VALIDATE_URL) === false) {
            return false;
        }
        $scheme = parse_url($url, PHP_URL_SCHEME);
        return $scheme === 'http' || $scheme === 'https';
    }

    /**
     * @param string $allowedChars a character-class body, e.g. "a-zA-Z0-9_-"
     */
    public static function isAllowListed(string $value, string $allowedChars): bool
    {
        return (bool) preg_match('/^[' . $allowedChars . ']*$/', $value);
    }

    /**
     * Guards against path traversal: rejects any path separators, NUL bytes,
     * and dot-only names. Returns the filename unmodified when safe.
     *
     * @throws \InvalidArgumentException on unsafe input
     */
    public static function sanitizeFilename(string $filename): string
    {
        if ($filename === '' || str_contains($filename, "\0")
            || str_contains($filename, '/') || str_contains($filename, '\\')) {
            throw new \InvalidArgumentException('securekit: invalid filename');
        }
        $trimmed = trim($filename);
        if ($trimmed === '' || $trimmed === '.' || $trimmed === '..') {
            throw new \InvalidArgumentException('securekit: invalid filename');
        }
        return $trimmed;
    }

    /**
     * HTML-escapes for safe inclusion in HTML body context (prevents XSS).
     */
    public static function escapeHtml(string $value): string
    {
        return htmlspecialchars($value, ENT_QUOTES | ENT_HTML5, 'UTF-8');
    }
}
